gcloud/datastore/test_client.py

```diff
@@ -18,29 +18,44 @@
 class TestClient(unittest2.TestCase):
 
     DATASET_ID = 'DATASET'
+    CONNECTION = object()
 
     def _getTargetClass(self):
         from gcloud.datastore.client import Client
         return Client
 
-    def _makeOne(self, dataset_id=DATASET_ID, namespace=None, connection=None):
-        return self._getTargetClass()(dataset_id, namespace=namespace,
-                                      connection=connection)
+    def _makeOne(self, dataset_id=DATASET_ID, connection=CONNECTION,
+                 namespace=None):
+        return self._getTargetClass()(dataset_id, connection=connection,
+                                      namespace=namespace)
 
-    def test_ctor_w_dataset_id_None(self):
-        self.assertRaises(ValueError, self._makeOne, None)
+    def test_ctor_w_dataset_id_no_environ(self):
+        self.assertRaises(EnvironmentError, self._makeOne, None)
 
-    def test_ctor_w_dataset_id_no_connection(self):
-        client = self._makeOne()
-        self.assertEqual(client.dataset_id, self.DATASET_ID)
+    def test_ctor_w_implicit_inputs(self):
+        from gcloud._testing import _Monkey
+        from gcloud.datastore import client as _MUT
+        OTHER = 'other'
+        conn = object()
+        klass = self._getTargetClass()
+        with _Monkey(_MUT,
+                     _determine_default_dataset_id=lambda x: x or OTHER,
+                     get_connection=lambda: conn):
+            client = klass()
+        self.assertEqual(client.dataset_id, OTHER)
+        self.assertEqual(client.namespace, None)
+        self.assertTrue(client.connection is conn)
 
     def test_ctor_w_explicit_inputs(self):
+        OTHER = 'other'
+        NAMESPACE = 'namespace'
         conn = object()
-        namespace = object()
-        client = self._makeOne(namespace=namespace, connection=conn)
-        self.assertEqual(client.dataset_id, self.DATASET_ID)
+        client = self._makeOne(dataset_id=OTHER,
+                               namespace=NAMESPACE,
+                               connection=conn)
+        self.assertEqual(client.dataset_id, OTHER)
+        self.assertEqual(client.namespace, NAMESPACE)
```

Comment:
I'm not sure where it comes from, but I don't like an `is` test for scalar values (it seems too "strong" an assertion, given that interning etc. isn't guaranteed anywhere).